Implement non-visible pipeline and iframe visibility methods

jmr0 opened the pull request:
This addresses #9566 and a good part of #9751, specifically:
- Pipeline has a notion of visibility
- IFrame setVisible/getVisible interface with IFrame's pipeline visibility
- IFrame mozbrowservisibilitychange responds to changes in visibility
- Pipeline visibility is used to limit animations (requestAnimationFrame does not tick animations when hidden) and to increase timer intervals (currently set to a minimum of 1 second while hidden)

Absent for now are any changes to the Document API and general implementation of the Page Visibility API, since the more interesting parts require knowledge of whether the user agent is minimized, OS screen locked, etc.

cc @paulrouget @jdm

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/10225)

<!-- Reviewable:end -->

jmr0 committed:
Implement non-visible pipeline and iframe visibility methods

highfive commented:
Heads up! This PR modifies the following files:
 * @jgraham: tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, tests/wpt/mozilla/meta/MANIFEST.json
 * @kichjang: components/script/dom/htmliframeelement.rs, components/script/timers.rs, components/script/dom/webidls/BrowserElement.webidl, components/script/dom/window.rs, components/script_traits/script_msg.rs, components/script_traits/script_msg.rs, components/script/script_thread.rs, components/script_traits/lib.rs, components/script_traits/lib.rs

larsbergstrom commented:
r? @jdm

emilio commented:
Just left a few nits and questions, @jdm should still review it thoroughly.

---

Review status: 0 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[components/compositing/pipeline.rs, line 363 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtytfvItGX8Q8iepDE:-KDtytfvItGX8Q8iepDF:-1472283869) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/compositing/pipeline.rs#L363)):*
nit: `} else {`

---

*[components/compositing/pipeline.rs, line 368 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtz72e01hgBl3b7gH3:-KDtz72e01hgBl3b7gH4:84945080) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/compositing/pipeline.rs#L368)):*
nit: Maybe a missing `unwrap()`?

---

*[components/compositing/pipeline.rs, line 369 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtz-KvgFxjuhvbxGXT:-KDtz-KvgFxjuhvbxGXU:467291472) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/compositing/pipeline.rs#L369)):*
nit: This `return` is not needed

---

*[components/script/script_thread.rs, line 1384 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtz8wj9VTr85lm8eBE:-KDtz8wj9VTr85lm8eBF:-1472283869) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/script/script_thread.rs#L1384)):*
nit: `} else {`

---

*[components/script/script_thread.rs, line 1397 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtzFDHznp8_clsiyw-:-KDtzFDHznp8_clsiyw0:-1800239257) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/script/script_thread.rs#L1397)):*
I'm not sure if we should panic here, there is ongoing work preventing constellation panics. Maybe a `warn!()` or `error!()` is enough?

---

*[components/script/dom/htmliframeelement.rs, line 228 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtydqHrYmNnIsZjBQ8:-KDtydqHrYmNnIsZjBQ9:421184989) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/script/dom/htmliframeelement.rs#L228)):*
Maybe only dispatch VisibilityChange when `visibility != self.visibility.get()`?

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 34 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtz_8u1kRcG6NaS-DJ:-KDtz_8u1kRcG6NaS-DK:503452439) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L34)):*
nit: whitespace

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10225#-:-KDtzbnOD4n1U95WlhVy:-1122268837)*
<!-- Sent from Reviewable.io -->

emilio commented:
Review status: 0 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[components/script/script_thread.rs, line 1397 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtzFDHznp8_clsiyw-:-KDu-RxbGU1vbU3QM5mz:-410691153) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/script/script_thread.rs#L1397)):*
Well, disregard the constellation part (this is script), but I still think we might not want to panic here. @jdm rules here :P

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10225)*
<!-- Sent from Reviewable.io -->

jmr0 commented:
*[components/script/dom/htmliframeelement.rs, line 228 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtydqHrYmNnIsZjBQ8:-KDu42QNaRQrTQvNvN2y:142865648) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/script/dom/htmliframeelement.rs#L228)):*
This turns out to be the case in practice because of a previous check in pipeline.rs, but probably best to have that check here explicitly either way

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10225)*
<!-- Sent from Reviewable.io -->

emilio commented:
Review status: 0 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[components/script/dom/htmliframeelement.rs, line 228 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtydqHrYmNnIsZjBQ8:-KDu72nx8U39wv3dCYXi:-1317727157) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/script/dom/htmliframeelement.rs#L228)):*
Sounds good, either that or a comment + `debug_assert!()` to make it explicit while reading this code could work :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10225)*
<!-- Sent from Reviewable.io -->

paulrouget commented:
I haven't tested, but can you make sure:
1) if `requestAnimationFrame` is called while not visible, it's properly called once visibility is set back to true?
2) visibility is recursively set in the frame tree (iframes inside iframes)

jmr0 commented:
1) This is the case now -- setting it back to visible results in compositing / running callbacks
2) Will do

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8641) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget commented:
@jmr0 not sure you saw what I was saying on IRC:

> <paul> jmr0: visibility is not a mozbrowser-only thing. A mozbrowser iframe happens to have methods to change the visibility, but the 'visibilitychange' event and the fact the page is slowed down si not specific the mozbrowsers.
> <paul> jmr0: if mozBrowserIframe.setVisible(true) is called, the mozbrowser iframe node element will fire the BrowserAPI-event mozbrowservisbilitychange, and all the documents within this iframe will fire a w3c-event 'visibilitychange'.

jmr0 commented:
Had a chat with @paulrouget on IRC and decided it might be better to work on the W3C Page Visibility API changes as part of a separate PR, noting that the visibility changes triggered by the Mozbrowser API should be reflected by documents within the Mozbrowser iframe (i.e. point #2 discussed above)

jmr0 commented:
Addressed @emilio's and @paulrouget's comments

jdm commented:
I'm really sorry that I haven't reviewed this yet; I've had a million other items vying for my attention. If I don't think I'm going to get to it this week I'll hand it off to somebody else.

paulrouget commented on components/script/script_thread.rs:
```diff
@@ -1370,6 +1377,51 @@ impl ScriptThread {
+                }
+                let document = inner_page.document();
+                for iframe in &document.collect_descendant_iframes() {
+                    iframe.set_visible(visible);
```
What about doing that at the pipeline level instead, where the frame tree is available.

paulrouget replied:
I mean in the constellation.

paulrouget replied:
Also - a frame can have multiple pipelines (history). And you want to change the visibility of these pipelines as well (in case of navigation). So this definitively needs to happen at the constellation level.

paulrouget commented on components/script/timers.rs:
```diff
@@ -211,6 +211,14 @@ impl OneshotTimers {
     }
 
+    pub fn slow_down(&self) {
+        self.js_timers.set_min_duration(MsDuration::new(1000)); //TODO: jmr0: where to store this constant?
```
Maybe in `resources/prefs.json`?

paulrouget commented on tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html:
```diff
@@ -0,0 +1,36 @@
+                t.done();
+            }
+        });
+
```
Can we also test `requestAnimationFrame` and `setTimeout` behavior?
Also test that subiframes are correctly affected?

paulrouget commented on components/script/script_thread.rs:
```diff
@@ -146,6 +146,8 @@ struct InProgressLoad {
     /// Window is frozen (navigated away while loading for example).
     is_frozen: bool,
+    /// Window is not visible.
+    is_hidden: bool,
```
Nit. Maybe you want to name that `is_visible`, like the other properties added to pipeline, pipeline_details and iframe .

paulrouget commented on components/script/script_thread.rs:
```diff
@@ -1370,6 +1377,51 @@ impl ScriptThread {
+            }
+        } else {
+            return;
+        }
```
Why don't you do that in `change_loaded_page_visibility`?

paulrouget commented:
I looked quickly at the PR and added some comments. Please note that I'm not a reviewer, so take my comments with a grain of salt :D

jdm commented:
Note: I'm about halfway through reviewing these changes.

jdm commented:
Thank you so much for tackling this, @jmr0! This is really great work, and I apologize for the delay in reviewing it. I think the general design here is correct, but there are some subtleties about what we can skip and where we should be sending messages that need to be sorted out. I also agree with the majority of the comments that @paulrouget left. Let us know if there's anything we can do to help you!

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 2 of 14 files at r1, 6 of 10 files at r2, 6 of 7 files at r3.
Review status: all files reviewed at latest revision, 26 unresolved discussions.

---

*[components/compositing/compositor.rs, line 255 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5LZ4ceDDekwR9YaUZ:-KF5LZ4ceDDekwR9YaU_:2142079174) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/compositor.rs#L255)):*
s/`current`/`this`, since an instance of `PipelineDetails` never refers to more than one pipeline.

---

*[components/compositing/compositor.rs, line 728 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF2abii5AlIgpiIA4K:-KFF2abjBuKwM-eG9xJB:1566048696) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/compositor.rs#L728)):*
I'm worried about what happens if an animation starts while a pipeline's invisible, and then the pipeline is made visible. I suspect the animation doesn't run.

---

*[components/compositing/compositor.rs, line 736 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF2p-EQuqz1q2GnbuG:-KFF2p-EQuqz1q2GnbuH:-1778885788) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/compositor.rs#L736)):*
Same concern about above with respect to requestAnimationFrame not getting called.

---

*[components/compositing/compositor.rs, line 1632 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5bQqYTsN78_1erv0-:-KF5bQqYTsN78_1erv00:-2131795961) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/compositor.rs#L1632)):*
nit: indent this line by one space.

---

*[components/compositing/constellation.rs, line 988 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF3eZVnLszA2-9U1es:-KFF3eZVnLszA2-9U1et:-1964682708) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/constellation.rs#L988)):*
This should probably be part of `new_pipeline` instead, to ensure that we never forget to propagate visibility to new child pipelines.

---

*[components/compositing/constellation.rs, line 993 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5jjv5FastvqbtAFBo:-KF5jjv5FastvqbtAFBp:1271727284) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/constellation.rs#L993)):*
Let's use `if let` instead of `map`.

---

*[components/compositing/constellation.rs, line 1353 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5jdku1yT2esTGAduK:-KF5jdkvWjoF8jJj5X3V:1271727284) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/constellation.rs#L1353)):*
Let's use `if let` instead of `map`.

---

*[components/compositing/pipeline.rs, line 368 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtz72e01hgBl3b7gH3:-KFF4X1Cre-eecZbBH2Z:-1451367163) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/compositing/pipeline.rs#L368)):*
Error check instead per #10423.

---

*[components/compositing/pipeline.rs, line 62 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF3raipNW3hSdcFIM9:-KFF3rajTOkzeTF3ZjGs:1115178979) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/pipeline.rs#L62)):*
Let's go with `/// Whether this pipeline should be treated as visible for the purposes of scheduling and resource management.`

---

*[components/compositing/pipeline.rs, line 361 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5k09AG_4itph9B9t1:-KF5k09AG_4itph9B9t2:824314310) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/pipeline.rs#L361)):*
We don't appear to use this return value anywhere.

---

*[components/compositing/pipeline.rs, line 366 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF3L1x7HRwsL27Vp5V:-KFF3L1x7HRwsL27Vp5W:1522262355) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/pipeline.rs#L366)):*
We should be marking all the child pipelines of this one similarly, since it's an inherited property.

---

*[components/compositing/pipeline.rs, line 368 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF4PooxquaenFiQQPs:-KFF4PooxquaenFiQQPt:-1501233662) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/pipeline.rs#L368)):*
We'll need to perform error checking here instead of calling unwrap, per #10423.

---

*[components/compositing/pipeline.rs, line 372 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFFDQCYlqDqr_7_Ri7J:-KFFDQCYlqDqr_7_Ri7K:549898098) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/pipeline.rs#L372)):*
The pipeline corresponding to `containing_id` could live in a separate script thread than the pipeline for `id`, so we'll need a separate message that is sent to that thread that is in charge of dispatching the mozbrowser event.

---

*[components/script/script_thread.rs, line 1397 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtzFDHznp8_clsiyw-:-KFF8gmBdYMO0oXeF8eY:1915050286) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/script/script_thread.rs#L1397)):*
Let's make this a warn!

---

*[components/script/script_thread.rs, line 1420 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtWFUg8QayLH7yYHva-r3-1420:-KFF9BWS4W6iyi5Z6_cL:-486773733) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/script/script_thread.rs#L1420)):*
I think the current code organization makes sense, since the other caller of change_loaded_page_visibility will never deal with the incomplete loads.

---

*[components/script/dom/document.rs, line 1495 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF9juX_NU5pu49OmM-:-KFF9juX_NU5pu49OmM0:1401290509) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/script/dom/document.rs#L1495)):*
We can remove this if we move all the child pipeline-related code to the constellation.

---

*[components/script/dom/htmliframeelement.rs, line 247 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFFCS2RsYQXpQJ3WYpa:-KFFCS2Sq44P6HI66t2L:721845508) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/script/dom/htmliframeelement.rs#L247)):*
Let's use `window.pipeline()` instead.

---

*[components/script/dom/webidls/BrowserElement.webidl, line 87 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5kwETDTLDVokAy3Bz:-KF5kwEUurogDhPWvwJu:1298708449) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/script/dom/webidls/BrowserElement.webidl#L87)):*
This should be `dom.mozbrowser.enabled` for future compatibility with #10081.

---

*[components/script/dom/webidls/BrowserElement.webidl, line 92 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5l5L7IRUr2k26OD_m:-KF5l5L7IRUr2k26OD_n:160841994) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/script/dom/webidls/BrowserElement.webidl#L92)):*
Same as above.

---

*[components/script_traits/lib.rs, line 127 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5kBQw92Dtge8nOh2J:-KF5kBQxu-67uMwGF0XB:1995685055) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/script_traits/lib.rs#L127)):*
I think merging these into one `ChangeFrameVisibilityStatus(/* parent */ PipelineId, /* frame */ PipelineId, bool)` message would make sense.

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 19 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFFF1mLP8bG3B6GSzr0:-KFFF1mMiloCbp0-QuCI:1960858897) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L19)):*
This can use `t.step_func`.

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 24 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFFF6-fCGahP47wLmZG:-KFFF6-fCGahP47wLmZH:168889165) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L24)):*
This should also use `t.step_func`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10225#-:-KFFFE9yNg3iOpPX5l4G:-1861742574)*
<!-- Sent from Reviewable.io -->

jmr0 commented:
Thanks @paulrouget and @jdm for taking a close look -- I'll go over these this weekend and let you know how I fared.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #9968) made this pull request unmergeable. Please resolve the merge conflicts.

jmr0 commented:
Sorry @jdm, things got pretty busy over on my end last week but I'm gonna try and wrap this up this upcoming week or let you know otherwise in case anyone else wants to take over

jmr0 commented:
Review status: 1 of 13 files reviewed at latest revision, 25 unresolved discussions.

---

*[components/compositing/compositor.rs, line 255 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5LZ4ceDDekwR9YaUZ:-KHBZHlsRCQ_MD2WaviL:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/compositor.rs#L255)):*
Done.

---

*[components/compositing/compositor.rs, line 728 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF2abii5AlIgpiIA4K:-KHItwdNV4L2wAfUwXob:-517518193) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/compositor.rs#L728)):*
I tested this scenario and it seemed to be OK -- the animation ends up "starting" at whatever step requestAnimationFrame is at, since that function continues to be called recursively (even for a non-visible pipeline) with some regularity and will reach this block with a "visible" state if pipeline visibility is set back to true. Is there a path in the code that makes suspect this wouldn't necessarily happen?

---

*[components/compositing/compositor.rs, line 1632 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5bQqYTsN78_1erv0-:-KHBZfuXuE3K-4gQP5RL:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/compositor.rs#L1632)):*
Done.

---

*[components/compositing/constellation.rs, line 988 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF3eZVnLszA2-9U1es:-KHB_BukreXYYFJOL8FO:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/constellation.rs#L988)):*
Done.

---

*[components/compositing/constellation.rs, line 993 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5jjv5FastvqbtAFBo:-KHGmnCGyui_VPSlZMEv:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/constellation.rs#L993)):*
Done.

---

*[components/compositing/constellation.rs, line 1353 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5jdku1yT2esTGAduK:-KHGkaIXCZw2MiT3RG2L:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/constellation.rs#L1353)):*
Done.

---

*[components/compositing/pipeline.rs, line 368 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtz72e01hgBl3b7gH3:-KHIxT8fgjWog27sBGpe:-477118135) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/compositing/pipeline.rs#L368)):*
I noticed this `compositor_proxy` didn't have unwrap() before

---

*[components/compositing/pipeline.rs, line 62 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF3raipNW3hSdcFIM9:-KHGm_9vZ0MpxRsl1S3Y:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/pipeline.rs#L62)):*
Done.

---

*[components/compositing/pipeline.rs, line 361 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5k09AG_4itph9B9t1:-KHGm11e9yOly8CGFFgW:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/pipeline.rs#L361)):*
Done.

---

*[components/compositing/pipeline.rs, line 366 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF3L1x7HRwsL27Vp5V:-KHGm1mxKwiJPw65QWdY:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/pipeline.rs#L366)):*
Done.

---

*[components/compositing/pipeline.rs, line 368 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF4PooxquaenFiQQPs:-KHGsX_8yUTDVcBrAi_-:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/pipeline.rs#L368)):*
Done.

---

*[components/compositing/pipeline.rs, line 372 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFFDQCYlqDqr_7_Ri7J:-KHIwfZKT0mf0RGWsAaB:416350853) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/pipeline.rs#L372)):*
This part I'm a little confused about. I was passing this along to the script thread so that it would be able to find pipeline `id` to then call `change_visibility_status`, as I noticed this wouldn't work otherwise. That being said, I wasn't exactly sure why this was the case

---

*[components/script/script_thread.rs, line 1397 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtzFDHznp8_clsiyw-:-KHHTRPL5WsY5W4XOpm6:-499146009) ([raw file](https://github.com/servo/servo/blob/09767e1f7fdd0ec3efaa02d69f4124115465408a/components/script/script_thread.rs#L1397)):*
Done.

---

*[components/script/script_thread.rs, line 150 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtWFUg8QayLH7yYHva-r3-150:-KHGsfsPiVogSgwhQyGI:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/script/script_thread.rs#L150)):*
Done.

---

*[components/script/script_thread.rs, line 1392 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KDtWFUg8QayLH7yYHva-r3-1392:-KHHtqieq8mK7G0OM-WE:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/script/script_thread.rs#L1392)):*
Done.

---

*[components/script/dom/webidls/BrowserElement.webidl, line 87 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5kwETDTLDVokAy3Bz:-KHHUFMkDprmNQLESkvM:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/script/dom/webidls/BrowserElement.webidl#L87)):*
Done.

---

*[components/script/dom/webidls/BrowserElement.webidl, line 92 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5l5L7IRUr2k26OD_m:-KHHUGJe_0aRkOw61u3f:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/script/dom/webidls/BrowserElement.webidl#L92)):*
Done.

---

*[components/script_traits/lib.rs, line 127 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KF5kBQw92Dtge8nOh2J:-KHHWHp1Q0vgTx3AbNq0:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/script_traits/lib.rs#L127)):*
Done.

---

*[components/script/dom/document.rs, line 1495 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF9juX_NU5pu49OmM-:-KHHWLlWQzJV_lyR_KZs:-499146009) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/script/dom/document.rs#L1495)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10225)*
<!-- Sent from Reviewable.io -->

jmr0 commented:
@jdm Still missing tests and a few minor changes, but I've replied to the comments I've addressed so far. Sorry it's taken so long! Inching closer I hope...

jdm commented:
-S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, jmr0 (Jose Rosello) wrote…</i></summary>
> @jdm Still missing tests and a few minor changes, but I've replied to the comments I've addressed so far. Sorry it's taken so long! Inching closer I hope...</details>


---

Reviewed 4 of 13 files at r4, 9 of 9 files at r5.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[components/compositing/compositor.rs, line 728 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF2abii5AlIgpiIA4K:-KHWlUqF2tWZMvfjPmgA:1210755957) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/compositor.rs#L728)):*
<details><summary><i>Previously, jmr0 (Jose Rosello) wrote…</i></summary>
> I tested this scenario and it seemed to be OK -- the animation ends up "starting" at whatever step requestAnimationFrame is at, since that function continues to be called recursively (even for a non-visible pipeline) with some regularity and will reach this block with a "visible" state if pipeline visibility is set back to true. Is there a path in the code that makes suspect this wouldn't necessarily happen?</details>
I was suspicious that we wouldn't do any compositing, but I guess it works!

---

*[components/compositing/pipeline.rs, line 372 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFFDQCYlqDqr_7_Ri7J:-KHWk_jNbS6LAz7FV_O1:-213014502) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/pipeline.rs#L372)):*
<details><summary><i>Previously, jmr0 (Jose Rosello) wrote…</i></summary>
> This part I'm a little confused about. I was passing this along to the script thread so that it would be able to find pipeline `id` to then call `change_visibility_status`, as I noticed this wouldn't work otherwise. That being said, I wasn't exactly sure why this was the case</details>
Imagine browser.html loads google.com inside a mozbrowser iframe. The iframe is contained in a document served from localhost, so we have two separate script threads: localhost and google.com. If we call `change_visibility` on the google.com pipeline, `self.script_chan` gives us a channel to the google.com script thread, but the iframe we care about lives in the localhost script thread instead.

---

*[components/compositing/pipeline.rs, line 220 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10225#-KHWc7nk34LeI7-BcKsr:-KHWc7nlPnojaQx-f83P:-254973030) ([raw file](https://github.com/servo/servo/blob/8eb82dad425b47443d3433d7d628daa105b72128/components/compositing/pipeline.rs#L220)):*
> ```Rust
>                                      state.parent_visibility.unwrap_or(true));
> 
>         pipeline.notify_visibility();
> ```

This should probably happen inside Pipeline::new if it's necessary.

---

*[components/compositing/pipeline.rs, line 405 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10225#-KHWhICoXv422YZ9I4Jj:-KHWhF-qb1uVCjMiqjrH:-1695811347) ([raw file](https://github.com/servo/servo/blob/8eb82dad425b47443d3433d7d628daa105b72128/components/compositing/pipeline.rs#L405)):*
> ```Rust
> 
>         self.script_chan.send(ConstellationControlMsg::ChangeFrameVisibilityStatus(parent_id, self.id, self.visible))
>                         .expect("Pipeline script chan");
> ```

We can use `let _ =` here instead of expect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10225#-:-KHWlj5WvswSJUSzsKVJ:1458266033)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #11044) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
@jmr0 Given the "work in progress" label on the most recent commit, do you want me to review this yet or hold off?

jmr0 commented:
@jdm ah yeah still missing some bits -- I'll flag you

jmr0 commented:
Review status: 0 of 14 files reviewed at latest revision, 12 unresolved discussions.

---

*[components/compositing/compositor.rs, line 728 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFF2abii5AlIgpiIA4K:-KIcNDhqPDzElHDKxVft:-1745678466) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/compositor.rs#L728)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> I was suspicious that we wouldn't do any compositing, but I guess it works!
</details>
So you were quite right to be suspicious! It seems like my original test had some other activity/animation going on that kept hitting this path (and thus made it seem like things were working, and I began to get suspicious myself). I've changed it and added a process_animations() call to the`PipelineVisibilityChanged`msg handler as well

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 38 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJ4GijGgzCih2Iy7nk6:-KJ4GijHJYsXMG1bxPxG:-1277072663) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L38)):*
> ```HTML
>     async_test(function(t) {
>       var iframe = document.createElement("iframe");
>       //var body = '<html><body><p>test</p></body></html>'; //TODO: jmr0 this doesn't work :(
> ```

That "helper.html" you see as src is simply this content, but I was unable to get a `contentWindow` object when I tried setting it as a data URI. Any thoughts on this?

---

*[components/compositing/pipeline.rs, line 372 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFFDQCYlqDqr_7_Ri7J:-KJ4EHeFqIIvX6XBUTtD:1532419699) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/components/compositing/pipeline.rs#L372)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> Imagine browser.html loads google.com inside a mozbrowser iframe. The iframe is contained in a document served from localhost, so we have two separate script threads: localhost and google.com. If we call `change_visibility` on the google.com pipeline, `self.script_chan` gives us a channel to the google.com script thread, but the iframe we care about lives in the localhost script thread instead.
</details>
That makes sense -- I decided to issue a `VisibilityChangeComplete` message from the inner frame's script thread back to the constellation to dispatch to whatever script thread is appropriate (it'd be itself for same domains). This seems to work well, but unsure as to how elegant it is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10225)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #11542) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
This is really close! Thanks for adding the tests, in particular!
-S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, bors-servo wrote…</i></summary>
> :umbrella: The latest upstream changes (presumably #11542) made this pull request unmergeable. Please resolve the merge conflicts.
</details>


---

Reviewed 5 of 14 files at r7, 10 of 10 files at r8.
Review status: all files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[components/constellation/constellation.rs, line 885 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl1fCWWWb-He6B8Nx2:-KJl1fCXxSqRNZSMBwlB:-1466770177) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/components/constellation/constellation.rs#L885)):*
> ```Rust
>             let parent_info = self.pipelines.get(&pipeline_id).and_then(|pipeline| pipeline.parent_info);
>             let window_size = self.pipelines.get(&pipeline_id).and_then(|pipeline| pipeline.size);
>             let parent_visibility = if let Some((parent_pipeline_id, _, _)) = parent_info {
> ```

I think this belongs in `new_pipeline` instead, rather than each caller.

---

*[components/constellation/constellation.rs, line 1469 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl21KmOfHBSbtpYhmP:-KJl21KmOfHBSbtpYhmQ:1596693420) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/components/constellation/constellation.rs#L1469)):*
> ```Rust
>         let mut child_pipeline_ids = vec!();
>         for frame in self.current_frame_tree_iter(frame_id) {
>             child_pipeline_ids.push(frame.current);
> ```

Why do we build the vector up separately?

---

*[components/constellation/constellation.rs, line 1472 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl26rtujqTztJZh1oC:-KJl26ru3yvzbwL7ntpY:2083578555) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/components/constellation/constellation.rs#L1472)):*
> ```Rust
>         }
>         for id in child_pipeline_ids {
>             self.pipelines.get_mut(&id).map(|pipeline| pipeline.change_visibility(visible));
> ```

Let's use `if let` here instead.

---

*[components/constellation/constellation.rs, line 1482 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl2Fg6LcGgGTME9Lmv:-KJl2Fg7Mmv2iRskIgg6:1196662102) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/components/constellation/constellation.rs#L1482)):*
> ```Rust
>                 parent_pipeline.script_chan.send(ConstellationControlMsg::NotifyVisibilityChange(parent_pipeline_id,
>                                                                                                  pipeline_id,
>                                                                                                  visibility)).unwrap();
> ```

We'll need to use the same error handling code code we do elsewhere in this file rather than calling `unwrap`.

---

*[components/constellation/pipeline.rs, line 407 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl2i2vjHFOspQ0O3f4:-KJl2i2vjHFOspQ0O3f5:1344493309) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/components/constellation/pipeline.rs#L407)):*
> ```Rust
>     pub fn change_visibility(&mut self, visible: bool) {
>         if visible == self.visible {
>             return; //No changes
> ```

I don't think this comment adds much :)

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 16 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl4gxH9Px5q9rj9m8t:-KJl4gxH9Px5q9rj9m8u:-1000892281) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L16)):*
> ```HTML
>         var iframe = document.createElement("iframe");
>         iframe.mozbrowser = true;
>         iframe.src = "http://www.mozilla.org";
> ```

This won't work; we want to be able to run tests without any network access and be confident about the actual content they run.

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 38 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJ4GijGgzCih2Iy7nk6:-KJl4tARBFevvj9pcnb4:612369868) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L38)):*
<details><summary><i>Previously, jmr0 (Jose Rosello) wrote…</i></summary>
> That "helper.html" you see as src is simply this content, but I was unable to get a `contentWindow` object when I tried setting it as a data URI. Any thoughts on this?
</details>
data: URLs are treated as cross-origin in servo right now. Having a same-origin HTML document is the way to go, so let's remove the commented lines.

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 44 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl54gyYEchKAUZBJ9R:-KJl54gzybAJeOK9Xlom:756006854) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L44)):*
> ```HTML
>       var start = null;
>       document.body.appendChild(iframe);
>       iframe.onload = function() {
> ```

`t.step_func(function() {`

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 50 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl58F7CbT-0lU9R6pY:-KJl58F7CbT-0lU9R6pZ:-79859000) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L50)):*
> ```HTML
>         element.style.position = 'relative';
>         element.style.right = "0px";
>         var step = function(timestamp) {
> ```

`t.step_func(function(timestamp) {`

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 61 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl5EHITk8Ux16c681K:-KJl5EHITk8Ux16c681L:1711377720) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L61)):*
> ```HTML
>         iframe.setVisible(false);
> 
>         iframe.contentWindow.setTimeout(function(){
> ```

t.step_func

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 66 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl5Ot3AXNiFFi09c6V:-KJl5Ot3AXNiFFi09c6W:1711377720) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L66)):*
> ```HTML
>         },1000);
> 
>         iframe.contentWindow.setTimeout(function(){
> ```

t.step_func

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 68 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl5RtZr5s9DnLKS4j2:-KJl5RtZr5s9DnLKS4j3:274681341) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L68)):*
> ```HTML
>         iframe.contentWindow.setTimeout(function(){
>             animationCompletesAfterResumingVisibility = element.style.right === '100px';
>             assert_equals(nonVisibleAnimationStopped && animationCompletesAfterResumingVisibility, true);
> ```

Let's have two separate `assert_true` calls here.

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 78 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl5kK6MKyun_cahJ2V:-KJl5kK7WpBjSFAsXYOg:1696674529) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L78)):*
> ```HTML
>     async_test(function(t) {
>       var iframe = document.createElement("iframe");
>       iframe.src = "http://web-platform.test:8000";
> ```

Let's use a particular helper html page for this (like making a `blank.html`).

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 81 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl5v1CLvdZMoY1vMUQ:-KJl5v1CLvdZMoY1vMUR:1711377720) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L81)):*
> ```HTML
>       iframe.mozbrowser = true;
>       iframe.onload = function() {
>           iframe.addEventListener("mozbrowservisibilitychange", e => {
> ```

t.step_func

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 83 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10225#-KJl5tnUQ7Z2IxFBteQx:-KJl5tnUQ7Z2IxFBteQy:1711377720) ([raw file](https://github.com/servo/servo/blob/463ee694a8eb0941c70e6d92814b6b042b447280/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L83)):*
> ```HTML
>           iframe.addEventListener("mozbrowservisibilitychange", e => {
>               var startTime = Date.now();
>               iframe.contentWindow.setTimeout(function() {
> ```

t.step_func

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10225#-:-KJl6AkeI2YsfPwBfKHv:890080791)*
<!-- Sent from Reviewable.io -->

jdm commented:
Only two outstanding issues remaining!
-S-needs-rebase -S-awaiting-review +S-needs-code-changes 

---

Reviewed 9 of 12 files at r9, 4 of 5 files at r10.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[components/constellation/constellation.rs, line 1521 \[r10\]](https://reviewable.io:443/reviews/servo/servo/10225#-KKG_wZCkmoygmcBc6vP:-KKG_wZCkmoygmcBc6vQ:-1746344659) ([raw file](https://github.com/servo/servo/blob/e34996900f55900d6c42024c7914e11bdbeb14d0/components/constellation/constellation.rs#L1521)):*
> ```Rust
>                                                                                      pipeline_id,
>                                                                                      visibility))
>                                .expect("Pipeline script chan");
> ```

We should be using `handle_send_error` here instead, like https://github.com/servo/servo/blob/5ec01d9dcafe81bfba3cd053015393baae5f3e37/components/constellation/constellation.rs#L713 .

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 19 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFFF1mLP8bG3B6GSzr0:-KKGahhJenkoYJVrBO6m:1205187691) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L19)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> This can use `t.step_func`.
</details>
This still needs to be addressed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10225#-:-KKGanogLaJKOuIoghJg:-1027182609)*
<!-- Sent from Reviewable.io -->

jmr0 commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 19 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFFF1mLP8bG3B6GSzr0:-KKGfCoB6TYFTIyILvPW:902688906) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L19)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> This still needs to be addressed.
</details>
Ah right -- I keep thinking these are only meant to wrap functions that interact with `t`. Could you remind me why we wrap all of these?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10225)*
<!-- Sent from Reviewable.io -->

jdm commented:
Thanks for sticking with this! Please squash these into one descriptive commit :)
-S-awaiting-review +S-needs-squash 

---

Reviewed 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html, line 19 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10225#-KFFF1mLP8bG3B6GSzr0:-KKLS-O1H4CjeK5Wq_GW:-276770699) ([raw file](https://github.com/servo/servo/blob/ac228c88dd1fd26c034689690e359b9652c34a9f/tests/wpt/mozilla/tests/mozilla/mozbrowser/iframe_visibility.html#L19)):*
<details><summary><i>Previously, jmr0 (Jose Rosello) wrote…</i></summary>
> Ah right -- I keep thinking these are only meant to wrap functions that interact with `t`. Could you remind me why we wrap all of these?
</details>
At minimum, it allows the test harness to notice any exceptions that might get thrown, so it's easier to debug if things go wrong. It's also required to asserts to function correctly inside of async_test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10225#-:-KKLSaabIc6pc7jYZ0Ag:-228308556)*
<!-- Sent from Reviewable.io -->

jmr0 commented:
@jdm Thanks for all your help -- definitely learned quite a bit more about Servo's architecture with this one :)

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 2bff131 has been approved by `jdm`

<!-- @bors-servo r=jdm 2bff131535987da0330737f78fb23291593db30d -->

bors-servo commented:
:hourglass: Testing commit 2bff131 with merge d620ab7...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1473), [arm32](http://build.servo.org/builders/arm32/builds/1582), [arm64](http://build.servo.org/builders/arm64/builds/1570), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1604), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1684), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1647), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1631), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1696), [windows](http://build.servo.org/builders/windows/builds/1305)

paulrouget commented:
Thank you @jmr0 ! This is excellent work.

jmr0 commented:
Thanks for all the help (and patience) @paulrouget! :)
